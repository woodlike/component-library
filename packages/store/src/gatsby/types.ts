import { NodeInput } from 'gatsby';
import { GraphQLOutputType } from 'graphql';

export interface ShopifyNode {
  readonly id: string;
  readonly title: string;
}

export interface ShopifyProductNode extends ShopifyNode {
  readonly description: string;
  readonly handle: string;
  readonly id: string;
  readonly images: ProductImage[];
  readonly fields: ProductFields;
  readonly features: ProductFeatures;
  readonly tags: string[];
  readonly title: string;
  readonly slug: string;
  readonly shopifyId: string;
  readonly variants: Variant[];
}

export interface ProductFeatures {
  readonly remoteId: string;
  readonly name: string;
  readonly title: string;
  readonly modelTitle: string;
  readonly features: string[];
  readonly fitAndCoverageTitle: string;
  readonly fitAndCoverageFeatureList: string[];
  readonly fabricFeature: FabricFeature;
  readonly productMarineProtection: ProductMarineProtection;
}

export interface FabricFeature {
  readonly remoteId: string;
  readonly title: string;
  readonly materialTitle: string;
  readonly description: string;
  readonly features: string[];
  readonly compositionTitle: string;
  readonly compositionFeatureList: string[];
}

export interface ProductMarineProtection {
  readonly remoteId: string;
  readonly title: string;
  readonly features: string[];
}

export interface ShopifyPageNode extends ShopifyNode {
  readonly body: string;
  readonly bodySummary: string;
}

export interface Variant {
  readonly available?: boolean;
  readonly compareAtPriceV2: Price | null;
  readonly compareAtLocalePrice: Price | null;
  readonly priceLocale: Price;
  readonly priceV2: Price;
  readonly shopifyId: string;
  readonly sku: string;
  readonly title: string;
  readonly weight: number;
  readonly weightUnit: string;
}

export interface Price {
  readonly amount: string;
  readonly currencyCode: string;
}

export interface ProductFields {
  readonly slug: string;
}

export interface ProductImage {
  readonly altText: string;
  readonly id: string;
  readonly originalSrc: string;
  readonly srcSet: SourceSetProps[];
  readonly localFile: {
    readonly name: string;
  };
}

export interface SourceSetProps {
  readonly src: string;
  readonly id?: string;
}

export interface GatsbyCtx<T> {
  readonly nodeModel: {
    getNodeById(args: {
      readonly id: string;
      readonly type?: string | GraphQLOutputType;
    }): (NodeInput & T[]) | null;
    getNodesByIds(args: {
      readonly ids: string[];
      readonly type?: string | GraphQLOutputType;
    }): (NodeInput & T[]) | null;
    getAllNodes(args: {
      readonly type: string | GraphQLOutputType;
    }): NodeInput & T[];
    getTypes(): NodeInput & T[];
  };
}

/**
 * The ProductImageSize are required for Shopify's url filter services that returns
 * optimized image sizes.
 */
export enum ProductImageSize {
  small = '500x500',
  medium = '800x800',
  large = '1024x1024',
}

export enum ProductImageWidth {
  small = 500,
  medium = 800,
  large = 1024,
}
