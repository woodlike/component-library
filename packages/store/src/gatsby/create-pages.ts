import { resolve } from 'path';
import { Actions, Reporter } from 'gatsby';
import { ShopifyProductNode } from '.';

export interface CreatePagesArgs {
  readonly actions: Actions;
  readonly reporter: Reporter;
  graphql: (query: string) => Promise<ProductQuery>;
}

interface ProductQuery {
  readonly data: ProductQueryData;
  readonly errors: boolean;
}

export interface ProductQueryData {
  readonly allShopifyProduct: {
    readonly edges: {
      readonly node: ShopifyProductNode;
    }[];
  };
}

export async function createPages({
  actions,
  graphql,
  reporter,
}: CreatePagesArgs): Promise<void> {
  const { createPage } = actions;
  try {
    const result: ProductQuery = await graphql(`
      query {
        allShopifyProduct {
          edges {
            node {
              id
              slug
            }
          }
        }
      }
    `);

    result.data.allShopifyProduct.edges.forEach(
      (edge: { readonly node: ShopifyProductNode }) => {
        createPage({
          path: edge.node.slug,
          component: resolve('./src/app/product/Layout.tsx'),
          context: {
            id: edge.node.id,
          },
        });
      },
    );
  } catch (error) {
    if (error) {
      reporter.panicOnBuild(`ðŸš¨  ERROR: creating Product pages: ${error}`);
      return Promise.reject(new Error(error));
    }
  }
}
