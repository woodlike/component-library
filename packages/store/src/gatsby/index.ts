export * from './create-pages';
export * from './create-slug';
export * from './resolvers';
export * from './schemas';
export * from './types';
