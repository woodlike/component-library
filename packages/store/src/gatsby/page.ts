import { resolve } from 'path';
import { Actions } from 'gatsby';

import { ShopifyProductNode, ProductNode } from '.';
import { Collection } from '../model';

export interface ProductQueryData {
  readonly allShopifyProduct: {
    readonly edges: ProductNode[];
  };
}

export interface CollectionQuery {
  readonly allShopifyCollection: {
    readonly nodes: Collection[];
  };
}

export function createProduct(edges: ProductNode[], actions: Actions) {
  const layerId = 'Shopify__Page__Z2lkOi8vc2hvcGlmeS9QYWdlLzEyNTg5NzAyOA==';
  const { createPage } = actions;
  edges.forEach((edge: { readonly node: ShopifyProductNode }) => {
    createPage({
      path: edge.node.slug,
      component: resolve('./src/app/product/Layout.tsx'),
      context: {
        id: edge.node.id,
        layerId,
      },
    });
  });
}

export function createCollection(nodes: Collection[], actions: Actions) {
  const { createPage } = actions;
  nodes.forEach(node => {
    createPage({
      path: node.slug,
      component: resolve('./src/app/collection/Layout.tsx'),
      context: {
        id: node.id,
      },
    });
  });
}
