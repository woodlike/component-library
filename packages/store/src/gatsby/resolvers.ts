import { Reporter } from 'gatsby';

import {
  GatsbyCtx,
  ProductFeatures,
  ProductImage,
  ProductImageSize,
  ShopifyProductNode,
  Variant,
} from '.';
import { priceFormatter } from '../utils';

const LOCALE = 'en-GB';

export function getNodeModelByRemoteId(
  type: string,
  id: string,
  ctx: GatsbyCtx<ProductFeatures>,
) {
  return (
    ctx.nodeModel.getAllNodes({ type }).find(node => node.remoteId === id) ??
    null
  );
}

export function productFeaturesResolver(
  source: ShopifyProductNode,
  _: object,
  ctx: GatsbyCtx<ProductFeatures>,
) {
  const product = ctx.nodeModel
    .getAllNodes({ type: 'GraphCMS_Product' })
    .find(product => {
      const regex = new RegExp(`(^\s*${product.title})`, 'gi');

      return regex.test(source.title);
    });

  if (!product) {
    return null;
  }

  const fabricFeature = getNodeModelByRemoteId(
    'GraphCMS_FabricFeature',
    product.fabricFeature.remoteId,
    ctx,
  );

  const productMarineProtection = getNodeModelByRemoteId(
    'GraphCMS_ProductMarineProtection',
    product.productMarineProtection.remoteId,
    ctx,
  );

  return {
    ...product,
    fabricFeature,
    productMarineProtection,
  };
}

export function shopifyProductResolver() {
  return {
    ShopifyProduct: {
      slug: {
        type: 'String!',
        resolve(source: ShopifyProductNode) {
          return `/products/${source.handle}`.replace(/\/\/+/g, '/');
        },
      },
      features: {
        resolve: productFeaturesResolver,
      },
    },
  };
}

export function shopifyProductImagesField() {
  return {
    ShopifyProductImages: {
      srcSet: {
        type: '[String!]!',
        resolve(source: ProductImage) {
          return Object.values(ProductImageSize).map(size =>
            source.originalSrc.replace(/(.jpg)+/gi, `_${size}.jpg`),
          );
        },
      },
    },
  };
}

export function productVariantPriceFields(reporter: Reporter) {
  return {
    ShopifyProductVariant: {
      compareAtLocalePrice: {
        type: 'ShopifyProductVariantPriceV2',
        resolve(source: Variant) {
          try {
            const { compareAtPriceV2 } = source;
            if (!!!compareAtPriceV2) {
              return null;
            }
            const { amount, currencyCode } = compareAtPriceV2;
            return {
              amount: priceFormatter(amount, LOCALE, currencyCode),
              currencyCode: currencyCode,
            };
          } catch (error) {
            reporter.panic(
              'ðŸš¨ Create Resolvers: Shopify product variant',
              error,
            );
          }
        },
      },
      priceLocale: {
        type: 'ShopifyProductVariantPriceV2!',
        resolve(source: Variant) {
          try {
            const {
              priceV2: { amount, currencyCode },
            } = source;

            return {
              amount: priceFormatter(amount, LOCALE, currencyCode),
              currencyCode: currencyCode,
            };
          } catch (error) {
            reporter.panic(
              'ðŸš¨ Create Resolvers: Shopify product variant',
              error,
            );
          }
        },
      },
    },
  };
}

export function createStoreResolvers(
  createResolvers: (schema: object) => void,
  reporter: Reporter,
): void {
  createResolvers({
    ...shopifyProductResolver(),
    ...shopifyProductImagesField(),
    ...productVariantPriceFields(reporter),
  });
}
