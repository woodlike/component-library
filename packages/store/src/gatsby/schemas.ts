import { Actions } from 'gatsby';

const schema = `
  type ShopifyCollection implements Node  {
    description: String!
    descriptionHtml: String!
    handle: String!
    id: String!
    image: Image
    shopifyId: String!
    slug: String!
    title: String!
    updatedAt: String
  }
  
  type ShopifyProduct implements Node  {
    id: String!
    description: String!
    images: [ShopifyProductImages!]!
    title: String!
    slug: String!
    variants: [ShopifyProductVariant!]!
    features: GraphCMS_Product
  }

  type ShopifyProductVariant implements Node @dontInfer {
    id: String!
    compareAtLocalePrice: ShopifyProductVariantPriceV2
    priceLocale: ShopifyProductVariantPriceV2!
    shopifyId: String!
    title: String!
  }

  type ShopifyProductVariantPriceV2 @dontInfer {
    amount: String!
    currencyCode: String!
  }

  type ShopifyProductImages @dontInfer {
    id: ID!
    altText: String
    originalSrc: String!
    srcSet: [String!]!
  }

  type Image @dontInfer {
    id: String!
    src: String!
  }
`;

export function createCustomSchema(actions: Actions): void {
  actions.createTypes(schema);
}
