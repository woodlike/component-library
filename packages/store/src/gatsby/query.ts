import { Collection } from "../model"
// import { NodeByLang } from "../model/page/utils"
import { ProductNode } from "."
import { ShopifyPage } from "../model/page"
// import { ShopifyPage } from "../model/page"

export interface CreatePageQuery {
  readonly data: {
    readonly allShopifyProduct: { readonly edges: ProductNode[] }
    readonly allShopifyCollection: { readonly nodes: Collection[] }
    readonly allLegalPageDe: ShopifyPage[]
    readonly allLegalPageEn: ShopifyPage[]
  }
  readonly errors: boolean
}

export async function create(
  graphql: (query: string) => Promise<CreatePageQuery>,
) {
  const { data }: CreatePageQuery = await graphql(`
    query {
      allShopifyProduct {
        edges {
          node {
            id
            slug
          }
        }
      }
      allShopifyCollection {
        nodes {
          id
          slug
        }
      }
      allLegalPageEn {
        title
        slug
        shopifyId
        id
        handle
        body
      }
      allLegalPageDe {
        title
        slug
        shopifyId
        id
        handle
        body
      }
    }
  `)
  return data
}
