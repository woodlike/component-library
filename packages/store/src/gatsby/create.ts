import * as GatsbyPage from "./page"
import * as Query from "./query"

import { Actions, Reporter } from "gatsby"

export interface CreatePagesArgs {
  readonly actions: Actions
  readonly reporter: Reporter
  graphql: (query: string) => Promise<Query.CreatePageQuery>
}

export async function createPages({
  actions,
  graphql,
  reporter,
}: CreatePagesArgs): Promise<void> {
  try {
    const data = await Query.create(graphql)

    const { edges } = data.allShopifyProduct
    const { nodes } = data.allShopifyCollection

    GatsbyPage.createCollection(nodes, actions)
    GatsbyPage.createLegal(data.allLegalPages, actions)
    GatsbyPage.createProduct(edges, actions)
  } catch (error) {
    if (error) {
      reporter.panicOnBuild(`ðŸš¨  ERROR: creating custom pages: ${error}`)
      return Promise.reject(new Error(error))
    }
  }
}
