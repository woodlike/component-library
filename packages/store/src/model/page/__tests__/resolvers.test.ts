import * as Page from ".."

import { ResolverArgs, ShopifyPage } from ".."

import { GatsbyCtx } from "../../../gatsby"
import { shopifyPages } from "../../../../__mocks__"

describe("Page Resolvers", () => {
  let resolverSrc: Record<string, string>
  let resolverArgs: ResolverArgs
  let resolverContext: GatsbyCtx<ShopifyPage>

  beforeEach(() => {
    resolverContext = Object.create({
      nodeModel: {
        getAllNodes: jest.fn().mockImplementation(() => shopifyPages.nodes),
      },
    })

    resolverSrc = {}
    resolverArgs = {}
  })

  afterEach(() => {
    jest.clearAllMocks()
    resolverSrc = (undefined as unknown) as Record<string, string>
    resolverArgs = (undefined as unknown) as ResolverArgs
    resolverContext = (undefined as unknown) as GatsbyCtx<ShopifyPage>
  })

  describe("Legal Pages", () => {
    it("should return an Array containing all the legal pages", () => {
      expect(
        Page.legal().allLegalPages.resolve(
          resolverSrc,
          resolverArgs,
          resolverContext,
        ),
      ).toMatchSnapshot()
    })
  })
})
