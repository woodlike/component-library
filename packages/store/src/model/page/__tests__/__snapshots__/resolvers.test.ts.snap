// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page Resolvers Legal Pages should return an Array containing all the legal pages in english 1`] = `
Array [
  Object {
    "handle": "terms-conditions",
    "id": "Shopify__Page__Z2lkOi8vc2hvcGlmeS9QYWdlLzEyOTAxNjkwMA==",
    "shortTitle": "Terms Conditions",
    "slug": "/legal/terms-conditions",
    "title": "Standard Business Terms and customer information",
  },
  Object {
    "handle": "privacy-policy",
    "id": "Shopify__Page__Z2lkOi8vc2hvcGlmeS9QYWdlLzEyOTIxNTc0OA==",
    "shortTitle": "Privacy Policy",
    "slug": "/legal/privacy-policy",
    "title": "Data protection declaration",
  },
  Object {
    "handle": "cancellation-right",
    "id": "Shopify__Page__Z2lkOi8vc2hvcGlmeS9QYWdlLzEzMDAwODc3Mg==",
    "shortTitle": "Cancellation Right",
    "slug": "/legal/cancellation-right",
    "title": "Returns & Exchanges",
  },
  Object {
    "handle": "imprint",
    "id": "Shopify__Page__Z2lkOi8vc2hvcGlmeS9QYWdlLzEzMDA2MTEyNA==",
    "shortTitle": "Imprint",
    "slug": "/legal/imprint",
    "title": "About Us - Imprint",
  },
]
`;

exports[`Page Resolvers Legal Pages should return an Array containing all the legal pages in german 1`] = `
Array [
  Object {
    "handle": "agb",
    "id": "Shopify__Page__Z2lkOi8vc2hvcGlmeS9QYWdlLzEyOTIxNTk0MA==",
    "shortTitle": "Agb",
    "slug": "/legal/agb",
    "title": "Allgemeine Geschäftsbedingungen und Kundeninformationen",
  },
  Object {
    "handle": "impressum",
    "id": "Shopify__Page__Z2lkOi8vc2hvcGlmeS9QYWdlLzEyOTIxNjA2OA==",
    "shortTitle": "Impressum",
    "slug": "/legal/impressum",
    "title": "Impressum",
  },
  Object {
    "handle": "datenschutzerklarung",
    "id": "Shopify__Page__Z2lkOi8vc2hvcGlmeS9QYWdlLzEyOTIxNjE5Ng==",
    "shortTitle": "Datenschutzerklarung",
    "slug": "/legal/datenschutzerklarung",
    "title": "Datenschutzerklärung",
  },
  Object {
    "handle": "widerrufsbelehrung",
    "id": "Shopify__Page__Z2lkOi8vc2hvcGlmeS9QYWdlLzI1MDc1MTMxMw==",
    "shortTitle": "Widerrufsbelehrung",
    "slug": "/legal/widerrufsbelehrung",
    "title": "Widerrufsbelehrung",
  },
]
`;
